AGENTS.md

```diff
@@ -0,0 +1,723 @@
+      "targetHandle": "par.data"
+    }
+  ],
+  "viewport": {"x": 0, "y": 0, "zoom": 1}
+}
+```
+
+### Path Prefixes in File References
+
+- `@/` - Relative to project directory
+- Absolute paths work as-is
+- URLs can reference remote resources
+
+## Migration System
+
+When schema changes occur, add migrations in `noodles-editor/src/noodles/__migrations__/`:
+
+```typescript
+export const migration = {
+  version: 7,
+  migrate(project: ProjectV6): ProjectV7 {
+    // Transform project structure
+    return transformedProject
+  }
+}
+```
+
+## Common Patterns
+
+### Accessing Operator Outputs
+
+```javascript
+// In CodeField or AccessorOp
+const data = op('/data-loader').out.data
+const threshold = op('./threshold').par.value
+```
+
+### Creating Layers
+
+```javascript
+// Operators that create Deck.gl layers should return LayerProps
+return {
+  type: ScatterplotLayer,
+  data: processedData,
+  getPosition: d => [d.lng, d.lat],
+  getRadius: 100,
+  getFillColor: [255, 0, 0]
+}
+```
+
+### Timeline Animation
+
+Any field can be keyframed:
+1. Fields are connected to Theatre.js via `useSheetValue` hook
+2. Changes in timeline propagate through reactive system
+3. Smooth interpolation between keyframes
+
+## Error Handling
+
+### Validation
+- Zod schemas validate all field values
+- Type mismatches caught at runtime
+- Clear error messages displayed in UI
+
+### Error Propagation
+```typescript
+try {
+  const result = operator.execute(inputs)
+  field.next(result)
+} catch (error) {
+  field.error(error)  // Propagate downstream
+}
+```
+
+### Debugging
+- Execution tracing tracks data flow
+- Performance profiling measures execution times
+- State inspection examines intermediate values
+
+## Best Practices
+
+### Graph Design
+- Minimize connections to reduce complexity
+- Group related operations in containers
+- Use descriptive node and field names
+- Document complex transformations
+
+### Performance
+- Avoid deep graph nesting
+- Batch related changes together
+- Profile and optimize hot paths
+- Use DuckDB for heavy data operations
+
+### Code Style
+- Follow Biome configuration
+- Write unit tests for operators
+- Use TypeScript strictly
+- Comment complex logic
+
+### Maintenance
+- Add migration scripts for schema changes
+- Version control graph changes
+- Keep documentation up-to-date
+- Test operators in isolation
+
+## Testing Strategy
+
+### When to Add Tests
+
+**Always add tests for:**
+- New operators and core functionality
+- Changes to critical components (listed below)
+- Complex state management or hook modifications
+- Bug fixes to prevent regressions
+- Non-trivial utility functions
+
+**Test Types:**
+
+- **Unit Tests**: For operator logic, pure functions, and utilities
+- **Integration Tests**: For graph transformations, hook interactions, and data flow
+- **Component Tests**: For React components with React Testing Library
+- **E2E Tests**: For full user workflows with Playwright
```

Comment:
Our current e2e test doesn't use playwright.. does it? Should it?

Could be good to add some guidance like we do for integration below such as

E2E

- prove the critical journeys work from “load project” to “inspect final flow output.”
- use Playwright sparingly for the most critical user journeys. Mouse/keyboard driven tests are inherently brittle. Prefer unit/integration/component tests for depth and coverage.